conformance_tests/test_tiovx/test_safe_casts.c

```diff
@@ -0,0 +1,624 @@
+    return status;
+}
+
+/* Check to see if the type is currently supported */
+vx_bool is_supported(vx_enum type)
+{
+    switch (type)
+    {
+        case VX_TYPE_ARRAY:
+        case VX_TYPE_CONTEXT:
+        case VX_TYPE_CONVOLUTION:
+        case VX_TYPE_DELAY:
+        case VX_TYPE_DISTRIBUTION:
+        case VX_TYPE_GRAPH:
+        case VX_TYPE_IMAGE:
+        case VX_TYPE_KERNEL:
+        case VX_TYPE_LUT:
+        case VX_TYPE_MATRIX:
+        case VX_TYPE_NODE:
+        case VX_TYPE_OBJECT_ARRAY:
+        case VX_TYPE_PARAMETER:
+        case VX_TYPE_PYRAMID:
+        case VX_TYPE_REMAP:
+        case VX_TYPE_SCALAR:
+        case VX_TYPE_TENSOR:
+        case VX_TYPE_THRESHOLD:
+        case VX_TYPE_USER_DATA_OBJECT:
+            return vx_true_e;
+        default:
+            return vx_false_e;
+   }
+}
+
+TEST(tivxSafeCasts, testCastGetFromSomeType)
+{
+    int i;
+    vx_graph graph = vxCreateGraph(context_->vx_context_);
+    ASSERT_VX_OBJECT(graph);
+    for (i = 0; i < UPCAST_NUM_TYPES; ++i)
+    {
+        vx_enum type = upcast_type_ids[i];
+        if (is_supported(type))
+        {
+            /* For both operations we are looking for the correct type and a reference count of 1, except for context, where ref count should be 2 */
+            vx_enum check_type = VX_ERROR_INVALID_TYPE;
+            vx_uint32 ref_count = 0;
+            vx_reference ref = castAsReference(graph, type);
+            ASSERT_VX_OBJECT(ref);
+            EXPECT_EQ_VX_STATUS(VX_SUCCESS, vxQueryReference(ref, VX_REFERENCE_TYPE, &check_type, sizeof(check_type)));
+            EXPECT_EQ_INT(check_type, type);
+            EXPECT_EQ_VX_STATUS(VX_SUCCESS, vxQueryReference(ref, VX_REFERENCE_COUNT, &ref_count, sizeof(ref_count)));
+            if (VX_TYPE_CONTEXT == type || VX_TYPE_KERNEL == type)
+            {
+                /* we expect both these to have an extra ref. count */
+                --ref_count;
+            }
+            EXPECT_EQ_INT(ref_count, 1U);
+            EXPECT_EQ_VX_STATUS(VX_SUCCESS, vxReleaseReference(&ref));
+            ref = getAsReference(graph, type);
+            ASSERT_VX_OBJECT(ref);
+            EXPECT_EQ_VX_STATUS(VX_SUCCESS, vxQueryReference(ref, VX_REFERENCE_TYPE, &check_type, sizeof(check_type)));
+            EXPECT_EQ_INT(check_type, type);
+            EXPECT_EQ_VX_STATUS(VX_SUCCESS, vxQueryReference(ref, VX_REFERENCE_COUNT, &ref_count, sizeof(ref_count)));
+            if (VX_TYPE_CONTEXT == type || VX_TYPE_KERNEL == type)
+            {
+                --ref_count;
+            }
+            EXPECT_EQ_INT(ref_count, 1U);
+            EXPECT_EQ_VX_STATUS(VX_SUCCESS, vxReleaseReference(&ref));
+        }
+    }
+    VX_CALL(vxReleaseGraph(&graph));
+}
+
+/* Check basic operation of castAs and GetAs */
+TEST(tivxSafeCasts, testCastGetAsSomeType)
+{
+    int i;
+    vx_graph graph = vxCreateGraph(context_->vx_context_);
+    ASSERT_VX_OBJECT(graph);)
```

Comment:
It looks like the ASSERT_VX_OBJECT requires two arguments, the second being the OpenVX type